odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,799 @@
+     [-0.5  0.5]]
+    [[ 0.5  0.5]
+     [ 0.5  0.5]]
+    [[-0.5  0.5]
+     [-0.5  0.5]]
+
+    References
+    ----------
+    .. _signal extension modes:
+       https://pywavelets.readthedocs.io/en/latest/ref/signal-extension-\
+modes.html
+    """
+    # Handle input
+    arr = np.asarray(arr)
+    wavelet = pywt_wavelet(wavelet)
+    mode, mode_in = str(mode).lower(), mode
+    if mode not in PYWT_SUPPORTED_MODES:
+        raise ValueError("mode '{}' not understood".format(mode_in))
+
+    # Compute single level DWT using pywt.dwtn and pick the approximation
+    # and detail coefficients from the dictionary
+    coeff_dict = pywt.dwtn(arr, wavelet, mode)
+    dict_keys = pywt_dict_keys(arr.ndim)
+    approx = coeff_dict[dict_keys[0]]
+    details = tuple(coeff_dict[k] for k in dict_keys[1:])
+    return approx, details
+
+
+def pywt_single_level_recon(approx, details, wavelet, mode, recon_shape=None):
+    """Return single level wavelet reconstruction from given coefficients.
+
+    Parameters
+    ----------
+    approx : `array-like`
+        Approximation coefficients.
+    details : sequence of `array-like`'s
+        Detail coefficients. The length of the sequence must be
+        ``2 ** ndim - 1``, where ``ndim`` is the number of dimensions
+        in ``approx``.
+    wavelet :  string or `pywt.Wavelet`
+        Specification of the wavelet to be used in the transform.
+        Use `pywt.wavelist` to get a list of available wavelets.
+    mode : string, optional
+        PyWavelets style signal extension mode. See `signal extension modes`_
+        for available options.
+    recon_shape : sequence of ints, optional
+        Shape of the array to be reconstructed. Without this parameter,
+        the reconstructed array always has even shape due to upsampling
+        by a factor of 2. To get reconstructions with odd shapes, this
+        parameter is required.
+
+    Returns
+    -------
+    recon : `numpy.ndarray`
+        The single-level wavelet reconstruction.
+
+    See Also
+    --------
+    pywt_single_level_decomp : Single-level decomposition, i.e. the
+        inverse of this function.
+    pywt_multi_level_recon : Multi-level version of the reconstruction.
+
+    Examples
+    --------
+    Take the coefficients from the example in `pywt_single_level_decomp`
+    and reconstruct the original array. Without ``recon_shape``, we get
+    a ``(4, 4)`` array as reconstruction:
+
+    >>> approx = [[1.5, 0.5],
+    ...           [0.5, 0.5]]
+    >>>
+    >>> details = ([[ 0.5, 0.5],
+    ...             [-0.5, 0.5]],
+    ...            [[ 0.5, 0.5],
+    ...             [ 0.5, 0.5]],
+    ...            [[-0.5, 0.5],
+    ...             [-0.5, 0.5]])
+    >>>
+    >>> # Gives even shape by default
+    >>> pywt_single_level_recon(approx, details, wavelet='haar', mode='zpd')
+    array([[ 1.,  1.,  1.,  0.],
+           [ 1.,  0.,  0.,  0.],
+           [ 0.,  1.,  1.,  0.],
+           [ 0.,  0.,  0.,  0.]])
+    >>>
```

Comment:
Blank line is perfectly fine by me You can even go blank line, comment, blank line. That gives a better style imo